### PR TITLE
Domain model changes required for SaaS integrations

### PR DESCRIPTION
**Pull request by nidhi-nair**
Needs additions for the jsonNode objects that will store the form schemas.
1. ~~Which of these can reuse existing params?~~ Fixed
2. ~~Do we store json configs in the db?~~ This will eventually be coming from cloud services

In a separate PR I'll be refactoring the execute action flow to make it more readable, it's LOC count is incredibly high at this point.

### CODE REVIEW

**Review by trishaanand — APPROVED** [review]

**Comment by trishaanand** [comment]
@nidhi-nair : Approved this. I may have missed it somehow, but I dont see any way to denote relationships between datasource templates and action templates. It might be possible in the future that some datasource templates support only certain action templates (depending on the authentication mechanism used), etc. Something to ponder about.

**Comment by nidhi-nair** [comment]
That's a fair ask. I've included this as a `TODO` so that @sumitsum can place the reference at the most suitable location when he takes up the changes required for datasources.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionTemplate.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An action template
```
Comment (sumitsum) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionTemplate.java:16]: Can you please add a comment explaining how it is meant to be used ?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionTemplate.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    private Map responseTransformationSpec;
+
+    private List<JsonNode> sampleResponses;
+
+    private String documentation; //Documentation for this particular API comes here
+
+    private String documentationUrl; //URL for this particular api's documentation comes here
+
+    private String versionId;
+}
```
Comment (sumitsum) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionTemplate.java:44]: Can you please add small comments explaining the fields: 
`template`
`requestTransformationSpec`
`responseTransformationSpec`
`sampleResponses`

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/DatasourceTemplate.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import com.appsmith.external.models.DatasourceConfiguration;
+import com.fasterxml.jackson.databind.JsonNode;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+/**
+ * A datasource template
```
Comment (sumitsum) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/domains/DatasourceTemplate.java:13]: Can you please add a comment explaining how it is meant to be used ?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/dtos/ActionDTO.java [hunk 4]
```diff
@@ -51,6 +51,8 @@
 
     ActionConfiguration actionConfiguration;
 
+    ActionConfiguration convertedActionConfiguration;
```
Comment (sumitsum) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/dtos/ActionDTO.java:54]: would it be ok if we rename this to `transformedActionConfiguration` ?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/NewActionServiceImpl.java [hunk 5]
```diff
@@ -548,6 +553,28 @@ public NewAction extractAndSetJsonPathKeys(NewAction newAction) {
                                                 actionConfiguration
                                         );
                                     })
+                                    .zipWith(actionMono)
+                                    .flatMap(tuple1 -> {
+                                        final ActionExecutionResult t1 = tuple1.getT1();
+                                        final NewAction t2 = tuple1.getT2();
```
Comment (sumitsum) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/NewActionServiceImpl.java:559]: Can we replace `t1` and `t2` with say `actionExecutionResult` and `action` - to improve the readability of the code that follows.

**Review by sumitsum — APPROVED** [review]

**Pull request merged**